### ISSUE TITLE
How to enable "determinism" or "reproducible" training?

### ISSUE THREAD

**Issue by Nic-Ma** [pytorch/ignite#786]
## ❓ Questions/Help/Support
Hi @vfdev-5 ,

1. We are trying to set random seeds for numpy, random and PyTorch, but I found you have logic to control them in ignite. Are they enough to make sure determinism training? How to enable and disable?
2. Should we use only 1 worker for PyTorch DataLoader?
3. Should we set "torch.backends.cudnn.benchmark=False"?

Thanks.

**Comment by vfdev-5** [comment]
Hi @Nic-Ma , 

I think it depends on what kind of determinism you wish to achieve... 
I think complete reproducibility up to 1e-10 tolerance on metrics and final model weights for every run can be tricky without any performance impact.
However, we can also admit some variation in final validation metrics and say that trainings are "reproducible". There are several aspects to keep in mind about this:
- software versions (which packages was used for the training)
- dataset versions (what dataset version was used for the training)
- configuration (model architecture, parameters, optimizer, criterion, hyperparameters, ...)
- randomness of dataflow (how and which the data is provided to compute gradients)
- randomness of cudnn ops (e.g. `torch.backends.cudnn.benchmark=True/False`)

First three points can be adressed by solutions like mlflow, polyaxon and others...
The last points can be partially controlled by user and ignite:

Ignite's engine by default sets random seed for each dataloader restart. This can not be disabled. This part intends to synchronize training data sampling indices and augmentations. When dataloader is `torch DataLoader` than we precompute batch indices in advance. Ultimately, we intended to be able to resume a crashed training from the last checkpoint (epoch or iteration). Resuming from a checkpoint that matches epoch (dataloader restart) matches the curves as there were no crash.  Resuming from a checkpoint somewhere in the middle of epoch (from an iteration) shows some gap due to non-synchronized data augmentations. For details, you can see [here](https://github.com/pytorch/ignite/tree/master/examples/contrib/cifar10#check-resume-training).

> Should we use only 1 worker for PyTorch DataLoader?

I think this is not necessary.

> Should we set "torch.backends.cudnn.benchmark=False"?

As [PyTorch docs](https://pytorch.org/docs/stable/notes/randomness.html) states about `torch.backends.cudnn.benchmark=False`, that the training speed is lower when `benchmark=False`. So, there is a tradeoff to deal with.

HTH

**Comment by Nic-Ma** [comment]
Hi @vfdev-5 ,

Thanks for your detailed explanation.
So you mean the normal ignite training(don't set seed manually and randomly) is already reproducible?
I will try to train several times on same dataset with workers=4 and compare the results.
BTW, how to **disable the logging output** from ignite? We have formatted log, want to disable ignite log.
Thanks.

**Comment by Nic-Ma** [comment]
Hi @vfdev-5 ,

BTW, we want to support a special metric that average other metrics, does it already supported in ignite?
For example, we computed 4 AUC for 4 labels in a multi-label task and want to use the average.
Should we manually compute the average AUC in user program?
Thanks.

**Comment by vfdev-5** [comment]
> BTW, how to disable the logging output from ignite? We have formatted log, want to disable ignite log.

@Nic-Ma , Engine is using standard logging, maybe passing Null handler can help

**Comment by Nic-Ma** [comment]
Hi @vfdev-5 ,

Thanks for your tips.
I didn't find API in engine to set handler. Do you mean set handler to engine.logger directly?
```py
engine.logger.addHandler(NullHandler())
```
Thanks.

**Comment by vfdev-5** [comment]
Yes, as it is a public attribute you can set it as you mentioned.
btw, I sent you an email on your publicly visible address at your github's profile (mp...@163.com), could you please take a look. Thanks.

**Comment by Nic-Ma** [comment]
Hi @vfdev-5 ,

Cool, thanks for that and also replied to your private email.
Thanks.

**Comment by vfdev-5** [comment]
I close the issue as answered. @Nic-Ma please, feel free to reopen it if needed more support on this subject.
